### DISCUSSION TITLE
LEGO Swerve drive

### DISCUSSION THREAD

**Discussion by thomasbrus** [pybricks/support#1729] [Show and tell]
### Four wheel swerve drive

Source code: https://github.com/thomasbrus/lego-swerve-drive

Well this was quite a long project 😮‍💨. I was set on building a swerve drive after [reading this article](https://dominik.win/blog/programming-swerve-drive/), and as far as I could tell there weren't many fully working ones on YouTube. At least none of them had simultaneous sidewards and rotational movement.

So my first attempt was building a four wheeled swerve drive. At some point I had all the code working, but in practice it turned out to be pretty difficult to make sure that all four wheels never ever drive in opposing directions turned. It was still working alright but you had to be very careful not to make sudden changes in direction.

The code is in the [four-wheel-swerve-drive branch](https://github.com/thomasbrus/lego-swerve-drive/tree/four-wheel-swerve-drive).  I built a dashboard in next js and using web bluetooth connected with the two hubs (since I had eight motors I needed two hubs). Then via the Gamepad API connected with the XBOX controller and send the joystick positions to the hubs, which then independently calculated the angle and speed for each two connected swerve drive wheels.

https://github.com/user-attachments/assets/bfb4f6e2-55bc-4f7a-9327-c3e47a821b10

![localhost_3000_ (1)](https://github.com/user-attachments/assets/b5249298-29e9-4da0-9e1a-0b1be4687859)

https://github.com/user-attachments/assets/e3e5413d-8507-44a3-9bab-0ea59a113be9

| image 1 | image 2 |
| ------ | ------ |
|![Omni 8-motors_5](https://github.com/user-attachments/assets/7ddd49e3-0e17-45b4-bedd-cb7389f513dc)|![Omni 8-motors_6](https://github.com/user-attachments/assets/f63170f7-4375-4ba3-ab04-74637b27340f)|

## Two wheel swerve drive

So then I switched to a two wheeled swerve drive with wheels that have less grip, and that worked quite well. The [main program](https://github.com/thomasbrus/lego-swerve-drive/blob/main/program/main.py) is fairly simple, I could reuse a lot of code from the four wheel swerve drive of course. There is also some code that simulates the Pybricks XBOXController class, which made debugging a lot easier.

Cool thing about this one is that it can rotate and move in any direction simultaneously, so mission accomplished 👌 

https://github.com/user-attachments/assets/cb462c00-1725-4b83-b89e-761eb8cfce54

| image 1 | image 2|
| -------- | ------- |
| ![IMG_1135](https://github.com/user-attachments/assets/c224a885-ab16-4b8c-97f7-3930c1ed2ee0) | ![IMG_1133](https://github.com/user-attachments/assets/46135a50-fafa-40fc-8ece-8549d89d9349) |

**Comment by laurensvalk** [comment]
This looks amazing! Thanks so much for sharing.

Do you think it might also work in this [kind of configuration](https://pybricks.com/project/technic-42146-xbox/) where the communication is `xbox controller` <---> `hub 1` <----> `hub 2`?

That's a really nice interface as well. On a somewhat related note, we're working on a dashboard-like interface here too https://github.com/pybricks/support/issues/1708 if you're interested in that sort of thing :)

**Reply by thomasbrus** [reply]
> This looks amazing! Thanks so much for sharing.

No problem, it's the least I can do, Pybricks is really cool 😀  

> Do you think it might also work in this [kind of configuration](https://pybricks.com/project/technic-42146-xbox/) where the communication is xbox controller <---> hub 1 <----> hub 2?

For sure that would work as well. I may use that set up for my next project! I'll have a look at the issue you linked. 

If I can give my two cents, for me it would be super helpful if the pybricksdev command was more reliable, often it freezes halfway when uploading a program. I tried turning off nearby bluetooth devices but it did not seem to make a difference. I don't know how well it works for other people, could not find much info about it. The longer the pybricks program the more trouble I had uploading it. Other than that, really great stuff, it's amazing that this exists.

**Reply by dlech** [reply]
If you want open an issue about the pybricksdev problem and include which OS you are using and a Bluetooth packet capture, we can look into it. (instructions https://bleak.readthedocs.io/en/latest/troubleshooting.html#capture-bluetooth-traffic)

**Reply by thomasbrus** [reply]
Well, I'm on Mac OS X (intel Sonoma 14.5), and it seems like the PacketLogger app isn't working at the moment. It's not capturing any bluetooth packages at all. I've used it a month or two ago and then it was working fine. 

I just installed the version that matches xcode, didn't touch anything, just kept the defaults 😄 There was a prompt for permission to capture bluetooth data, entered my password there, but no luck. I think this is the same issue: https://forums.developer.apple.com/forums/thread/759461.  I'll give it another go after the next mac update.

**Reply by thomasbrus** [reply]
See https://github.com/pybricks/pybricksdev/issues/94
